internal/memory/git.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 Lekko Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memory
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"buf.build/gen/go/lekkodev/cli/bufbuild/connect-go/lekko/backend/v1beta1/backendv1beta1connect"
+	backendv1beta1 "buf.build/gen/go/lekkodev/cli/protocolbuffers/go/lekko/backend/v1beta1"
+	"github.com/bufbuild/connect-go"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/cache"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/go-git/go-git/v5/storage/filesystem"
+	"github.com/rjeczalik/notify"
+	"google.golang.org/protobuf/proto"
+)
+
+// Constructs an in-memory store that fetches configs from a local git repo at the given path.
+// If api key is empty, the store runs in local (offline) mode, and does not communicate with Lekko.
+func NewGitStore(ctx context.Context, apikey, url, ownerName, repoName, path string) (Store, error) {
+	var distClient backendv1beta1connect.DistributionServiceClient
+	if len(apikey) > 0 {
+		distClient = backendv1beta1connect.NewDistributionServiceClient(http.DefaultClient, url)
+	}
+	fs := osfs.New(path)
+	gitfs, err := fs.Chroot(git.GitDirName)
+	if err != nil {
+		return nil, err
+	}
+	storer := filesystem.NewStorage(gitfs, cache.NewObjectLRUDefault())
+	return newGitStore(ctx, apikey, ownerName, repoName, storer, fs, distClient, eventsBatchSize, true)
+}
+
+func newGitStore(
+	ctx context.Context,
+	apikey, ownerName, repoName string,
+	storer storage.Storer, fs billy.Filesystem,
+	distClient backendv1beta1connect.DistributionServiceClient,
+	eventsBatchSize int, watch bool,
+) (*gitStore, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	g := &gitStore{
+		distClient: distClient,
+		store:      newStore(),
+		repoKey: &backendv1beta1.RepositoryKey{
+			OwnerName: ownerName,
+			RepoName:  repoName,
+		},
+		apikey: apikey,
+		cancel: cancel,
+		storer: storer,
+		fs:     fs,
+	}
+	if distClient != nil {
+		// register with lekko backend
+		sessionkey, err := g.registerWithBackoff(ctx)
+		if err != nil {
+			// silently fail on registration errors in git mode
+			log.Printf("error registering lekko client: %v", err)
+		}
+		g.sessionkey = sessionkey
+		g.eb = newEventBatcher(ctx, distClient, g.sessionkey, g.apikey, eventsBatchSize)
+	}
+	if _, err := g.load(ctx); err != nil {
```

Comment:
It's probably just a go convention that I need to get used to, but I still find it hard to track single letter names across the function :) I wonder if there is a style when "instance pointer" is called `this` or something similar :) ?

Reply:
the convention is generally single letter for pointer receivers, unless there's a conflict - https://go.dev/tour/methods/4

agreed that its not as clear as some other languages e.g. java/python with this/self